### ISSUE TITLE
Chopin tradebeans/tradesoap exits with error code when forced to use compilers

### ISSUE THREAD

**Issue by caizixian** [dacapobench/dacapobench#198]
Commit: 69a704e
```
/usr/lib/jvm/temurin-11-amd64/bin/java -version
openjdk version "11.0.12" 2021-07-20
OpenJDK Runtime Environment Temurin-11.0.12+7 (build 11.0.12+7)
OpenJDK 64-Bit Server VM Temurin-11.0.12+7 (build 11.0.12+7, mixed mode)
```

```
/usr/lib/jvm/temurin-11-amd64/bin/java -server -Xcomp -XX:+UseG1GC -Xms618M -Xmx618M -jar /usr/share/benchmarks/dacapo/dacapo-evaluation-git-69a704e.jar -n 5 tradesoap
--------------------------------------------------------------------------------
IMPORTANT NOTICE:  This is NOT a release build of the DaCapo suite.
Since it is not an official release of the DaCapo suite, care must be taken when
using the suite, and any use of the build must be sure to note that it is not an
offical release, and should note the relevant git hash.

Feedback is greatly appreciated.   The preferred mode of feedback is via github.
Please use our github page to create an issue or a pull request.
    https://github.com/dacapobench/dacapobench.
--------------------------------------------------------------------------------

Using scaled threading model. 24 processors detected, 24 threads used to drive the workload, in a possible range of [1,128]
Launching the server
```

The above exits with code 1. Removing `-Xcomp` seems to solve the problem.

**Comment by steveblackburn** [comment]
After upgrading the workloads, this problem remains.  Since the JVM dies silently, it's not straightforward to diagnose.

**Comment by steveblackburn** [comment]
See also https://bugs.openjdk.org/browse/JDK-8067708

**Comment by caizixian** [comment]
Filed an OpenJDK issue https://bugs.openjdk.org/browse/JDK-8316639

**Comment by caizixian** [comment]
The stacktrace points to a potential DaCapo problem. The JBS issue is closed for now.
```
j java.lang.Shutdown.halt0(I)V+0 java.base
j java.lang.Shutdown.halt(I)V+7 java.base
j java.lang.Runtime.halt(I)V+17 java.base
j org.dacapo.daytrader.DaCapoServerRunner.initialize()V+95
j java.lang.invoke.LambdaForm$DMH+0x00000007c0161800.invokeStatic(Ljava/lang/Object;)V+9 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c0162800.invoke(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+12 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c1400.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+19 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+41 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j org.dacapo.daytrader.Launcher.initialize(Ljava/io/File;Ljava/io/File;IIIZ)V+114
j java.lang.invoke.LambdaForm$DMH+0x00000007c0155c00.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;IIII)V+20 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c015b000.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+154 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7c00.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+104 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j org.dacapo.harness.Tradebeans.prepare(Ljava/lang/String;)V+259
j org.dacapo.harness.Benchmark.run(Lorg/dacapo/harness/Callback;Ljava/lang/String;)Z+20
j org.dacapo.harness.TestHarness.runBenchmark(Ljava/io/File;Ljava/io/File;Ljava/lang/String;Lorg/dacapo/harness/TestHarness;)V+109
j org.dacapo.harness.TestHarness.main([Ljava/lang/String;)V+720
j java.lang.invoke.LambdaForm$DMH+0x00000007c00c5400.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;)V+10 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7800.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+33 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7c00.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+55 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j Harness.main([Ljava/lang/String;)V+44
v ~StubRoutines::call_stub 0x00007f2efc417d21
V [libjvm.so+0xec85c6] JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, JavaThread*)+0x466 (javaCalls.cpp:415)
V [libjvm.so+0xff04ed] jni_invoke_static(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, JavaThread*) [clone .isra.0] [clone .constprop.1]+0x33d (jni.cpp:888)
V [libjvm.so+0xff38d4] jni_CallStaticVoidMethod+0x194 (jni.cpp:1712)
C [libjli.so+0x528e] JavaMain+0xc0e (java.c:547)
C [libjli.so+0x7fad] ThreadJavaMain+0xd (java_md.c:650)
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j java.lang.Shutdown.halt0(I)V+0 java.base
j java.lang.Shutdown.halt(I)V+7 java.base
j java.lang.Runtime.halt(I)V+17 java.base
j org.dacapo.daytrader.DaCapoServerRunner.initialize()V+95
j java.lang.invoke.LambdaForm$DMH+0x00000007c0161800.invokeStatic(Ljava/lang/Object;)V+9 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c0162800.invoke(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+12 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c1400.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+19 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+41 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j org.dacapo.daytrader.Launcher.initialize(Ljava/io/File;Ljava/io/File;IIIZ)V+114
j java.lang.invoke.LambdaForm$DMH+0x00000007c0155c00.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;IIII)V+20 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c015b000.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+154 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7c00.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+104 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j org.dacapo.harness.Tradebeans.prepare(Ljava/lang/String;)V+259
j org.dacapo.harness.Benchmark.run(Lorg/dacapo/harness/Callback;Ljava/lang/String;)Z+20
j org.dacapo.harness.TestHarness.runBenchmark(Ljava/io/File;Ljava/io/File;Ljava/lang/String;Lorg/dacapo/harness/TestHarness;)V+109
j org.dacapo.harness.TestHarness.main([Ljava/lang/String;)V+720
j java.lang.invoke.LambdaForm$DMH+0x00000007c00c5400.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;)V+10 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7800.invoke(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+33 java.base
j java.lang.invoke.LambdaForm$MH+0x00000007c00c7c00.invokeExact_MT(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+20 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+55 java.base
j jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+23 java.base
j java.lang.reflect.Method.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;+102 java.base
j Harness.main([Ljava/lang/String;)V+44
v ~StubRoutines::call_stub 0x00007f2efc417d21 
```

I think the culprit is https://github.com/dacapobench/dacapobench/blob/0f83db5d60e0362ca861084d8aa8713321a5e81b/benchmarks/libs/daytrader/src/org/dacapo/daytrader/DaCapoServerRunner.java#L39 which is very similar to the Kafka issue #226. 

I suspect that `-Xcomp` forces compilation, and the server might not full boot in 5 seconds due to ongoing JIT compilation. The DaCapo error messages were not properly printed probably due to the use of `Runtime.getRuntime().halt(1)`.
https://github.com/dacapobench/dacapobench/blob/0f83db5d60e0362ca861084d8aa8713321a5e81b/benchmarks/libs/daytrader/src/org/dacapo/daytrader/DaCapoServerRunner.java#L45

This also means that #225 could also be valid when running on a slower machine.

**Comment by steveblackburn** [comment]
This makes sense.   I am investigating a fix.  Thank you.

**Comment by steveblackburn** [comment]
I've fixed this issue in 25fda5afa7630afbf3304eedb44325eecacd85c3.

There were a number of problems.

1. When wildfly is started, stderr and stdout are redirected to the server log.  However, this is not done atomically, which means that messages sent to stderr for the first second or two after startup are lost (neither on console or wildfly's log).  This (and lack of awareness that this behavior existied) made debugging difficult.
2. If the client tries to start before the daytrader application is deployed, it will fail.  The original startup code would sleep the startup thread for a *fixed* period.
3. If exceptions (including connection exceptions, which are expected) occurred during startup then the benchmark would crash silently.
4. To avoid a runaway server, the daytrader harness has code that will kill it if it has not died before some time has elapsed.   That code was placed within each iteration and the timeout was short, so when many iterations were run the first iteration could cause the system to shut down.

The fix in 0a1d88b2 does a number of things.

- It uses a generic watchdog timer 7c12daff395b75b374cf9c4b23c2cf6bc86f3934 to catch non-termination (which should never happen, but is not uncommon when debugging a GC or compiler).
  - This avoids an ad hoc solution
  - Because stderr is not redirected in the generic watchdog, error messages appear on the console
  - The generic watchdog provides info on the watchdog and how to adjust the timeout (which is generous by default)
- To provide suitable defaults, we now include nominal benchmark times in the cnf files which are used to set the watchdog
- The startup code now loops at a 200ms period checking for the server.  The loop is infinite, but the watchdog will take care of a runaway.
